beacon-chain/blockchain/service.go

```diff
@@ -206,3 +211,22 @@ func (c *ChainService) ChainHeadRoot() ([32]byte, error) {
 	}
 	return root, nil
 }
+
+// IsCanonical returns true if the input block hash of the corresponding slot
+// is part of the canonical chain. False otherwise.
+func (c *ChainService) IsCanonical(slot uint64, hash []byte) bool {
+	c.canonicalBlocksLock.RLock()
+	defer c.canonicalBlocksLock.RUnlock()
+	if canonicalHash, ok := c.canonicalBlocks[slot]; ok {
+		return bytes.Equal(canonicalHash, hash)
+	}
+	return false
+}
+
+// InsertsCanonical inserts a canonical block hash to its corresponding slot.
+// This is used for testing purpose.
+func (c *ChainService) InsertsCanonical(slot uint64, hash []byte) {
```

Comment:
Please remove this method. We shouldn't have methods like this unless absolutely necessary. It's also confusing that there is the duplicated logic in fork_choice.go

Reply:
Will remove in a subsequent PR